op-e2e/actions/op_program_test.go

```diff
@@ -0,0 +1,270 @@
+	l2RPC := env.engine.RPCClient()
+	l2Client, err := host.NewL2Client(l2RPC, env.log, nil, &host.L2ClientConfig{L2ClientConfig: l2ClCfg, L2Head: l2Head})
+	require.NoError(t, err, "failed to create L2 client")
+	l2DebugCl := &host.L2Source{L2Client: l2Client, DebugClient: sources.NewDebugClient(l2RPC.CallContext)}
+	dfault.L2ProcessSource = l2DebugCl
+
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+func Test_ProgramAction_SimpleEmptyChain_HonestClaim_Granite(gt *testing.T) {
+	t := NewDefaultTesting(gt)
+	tp := NewTestParams()
+	dp := NewDeployParams(t, func(dp *e2eutils.DeployParams) {
+		genesisBlock := hexutil.Uint64(0)
+
+		// Enable Cancun on L1 & Granite on L2 at genesis
+		dp.DeployConfig.L1CancunTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisRegolithTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisCanyonTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisDeltaTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisEcotoneTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisFjordTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisGraniteTimeOffset = &genesisBlock
+	})
+	bCfg := NewBatcherCfg()
+	env := NewL2FaultProofEnv(t, tp, dp, bCfg)
+
+	// Build an empty block on L2
+	env.sequencer.ActL2StartBlock(t)
+	env.sequencer.ActL2EndBlock(t)
+
+	// Instruct the batcher to submit the block to L1, and include the transaction.
+	env.batcher.ActSubmitAll(t)
+	env.miner.ActL1StartBlock(12)(t)
+	env.miner.ActL1IncludeTxByHash(env.batcher.LastSubmitted.Hash())(t)
+	env.miner.ActL1EndBlock(t)
+
+	// Finalize the block with the batch on L1.
+	env.miner.ActL1SafeNext(t)
+	env.miner.ActL1FinalizeNext(t)
+
+	// Instruct the sequencer to derive the L2 chain from the data on L1 that the batcher just posted.
+	env.sequencer.ActL1HeadSignal(t)
+	env.sequencer.ActL2PipelineFull(t)
+
+	l1Head := env.miner.l1Chain.CurrentBlock()
+	l2SafeHead := env.engine.l2Chain.CurrentSafeBlock()
+
+	// Ensure there is only 1 block on L1.
+	require.Equal(t, uint64(1), l1Head.Number.Uint64())
+	// Ensure the block is marked as safe before we attempt to fault prove it.
+	require.Equal(t, uint64(1), l2SafeHead.Number.Uint64())
+
+	// Fetch the pre and post output roots for the fault proof.
+	preRoot, err := env.sequencer.RollupClient().OutputAtBlock(context.Background(), l2SafeHead.Number.Uint64()-1)
+	require.NoError(t, err)
+	claimRoot, err := env.sequencer.RollupClient().OutputAtBlock(context.Background(), l2SafeHead.Number.Uint64())
+	require.NoError(t, err)
+
+	// Run the fault proof program from the state transition from L2 block 0 -> 1.
+	programCfg := NewOpProgramCfg(
+		t,
+		env,
+		l1Head.Hash(),
+		preRoot.BlockRef.Hash,
+		common.Hash(preRoot.OutputRoot),
+		common.Hash(claimRoot.OutputRoot),
+		l2SafeHead.Number.Uint64(),
+	)
+	err = host.FaultProofProgram(context.Background(), env.log, programCfg)
+	require.NoError(t, err)
+}
+
+func Test_ProgramAction_SimpleEmptyChain_JunkClaim_Granite(gt *testing.T) {
```

Comment:
nit: this test is almost identical to the above. Could reuse the code using a generic `emptyChain` test that asserts on an input claim.

Reply:
Yeah for sure. These are mostly just examples. As we start to flesh out the action test suite, glad to modify - right now not sure if there will be more empty chain tests, etc.

Ideally we stick to more complex actions here, which will each (hopefully) have varying setups, forks activated, etc.

Reply:
ya makes sense. We can leave em like this as they're pretty illustrative.